extensions/amp-google-assistant-assistjs/0.1/assistjs-config-service.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview The shared service takes assist.js related config from <amp-google-assistant-assistjs-config>
+ * and provides config parameters to other assist.js components.
+ */
+
+import {Deferred} from '../../../src/utils/promise';
+import {hasOwn} from '../../../src/utils/object';
+
+export class AssistjsConfigService {
+  /** */
+  constructor() {
+    /** @private {string} */
+    this.projectId_ = '';
+
+    /** @private {string} */
+    this.assistjsServer_ = 'https://actions.google.com';
+
+    /** @private {boolean} */
+    this.devMode_ = false;
+
+    /** @private {string} */
+    this.hostUrl_ = window.location.href;
+
+    /** @private {Deferred} */
+    this.initializedDeferred_ = new Deferred();
+  }
+
+  /** @return {string} */
+  initializeConfigs() {
+    const configElements = document.getElementsByTagName(
```

Comment:
Why isn't this work done by the `amp-google-assistant-assistjs-config` element?

Reply:
The Config element gathers all the configuration needed by the other elements. Its a glorified JSON wrapper. Whats the best way of getting that config into each of the other elements?

Is it really better to have each of them find the config CE on the page and call getters? Or are you saying the config element would do this work and then call a setter on the service?

Reply:
The config element should be doing this before initializing the service. The other elements just need to get the already initialized service, then.

Reply:
Effectively, this initialization is triggered by the config element during its buildCallback(). You can see amp-google-assistant-assistjs.js for more details. 

I am not sure how we can finish this initialization before service got even constructed. My current understanding of the services are that they are constructed when document get built, so you don't construct them manually with some custom parameters. So the only way we can set the service state is by calling some setter functions, e.g. initializeConfigs(). Since  the service can already get the config element and its JSON to initialize the service, I move the entire initialization logic into this setter function. Anything I am missing here?

We can move the JSON extraction logic to buildCallback() of the config element, but we still need to call a setter function to set those states in ConfigService before it can be used by other elements. I don't have strong opinions on that. What do you guys think?

Reply:
I see. Thanks for the info.

Done.